burn-compute/src/tune/tune_cache.rs

```diff
@@ -20,24 +43,116 @@ pub enum TuneCacheResult<K> {
+
+        #[cfg(feature = "autotune-persistent-cache")]
+        {
+            if let Some((is_checked, index)) = result {
+                if !*is_checked {
+                    let checksum = autotune_operation_set.compute_checksum();
+                    let (expected_checksum, _) = self
+                        .persistent_cache
+                        .get(&key)
+                        .expect("Both caches should be in sync");
+                    if &checksum != expected_checksum {
+                        return TuneCacheResult::Miss(autotune_operation_set);
+                    }
+                    *is_checked = true;
+                }
+                return TuneCacheResult::Hit(autotune_operation_set.fastest(*index));
+            }
+        }
+
+        #[cfg(not(feature = "autotune-persistent-cache"))]
+        {
+            if let Some((_is_checked, index)) = result {
+                return TuneCacheResult::Hit(autotune_operation_set.fastest(*index));
+            }
         }
+
         TuneCacheResult::Miss(autotune_operation_set)
     }
 
     pub(crate) fn cache_insert(&mut self, key: K, fastest_index: usize) {
-        self.cache.insert(key, fastest_index);
+        self.cache.insert(key, (true, fastest_index));
+    }
+
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn persistent_cache_insert(
+        &mut self,
+        key: K,
+        checksum: String,
+        fastest_index: usize,
+    ) {
+        self.persistent_cache.insert(key, (checksum, fastest_index));
+    }
+
+    /// Load the persistent cache data from disk
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn load(&mut self) -> Result<(), io::Error> {
+        let file_path = get_persistent_cache_file_path();
+        // note: reading file ro memory is faster than using
+        // serde from_reader with a buffered reader
+        // see issue:
+        // https://github.com/serde-rs/json/issues/160
+        match fs::read_to_string(file_path) {
+            Ok(data) => {
+                let data: Vec<(K, (String, usize))> = serde_json::from_str(&data)?;
+                for (key, value) in data.into_iter() {
+                    self.persistent_cache.insert(key, value);
+                }
+                Ok(())
+            }
+            Err(e) => {
+                if e.kind() == std::io::ErrorKind::NotFound {
+                    Ok(())
+                } else {
+                    Err(e)
+                }
+            }
+        }?;
+        for (key, (_checksum, index)) in self.persistent_cache.iter() {
+            self.cache.insert(key.clone(), (false, *index));
+        }
+        Ok(())
+    }
+
+    /// Save the persistent cache on disk
+    #[cfg(feature = "autotune-persistent-cache")]
+    pub(crate) fn save(&self) {
+        let file_path = get_persistent_cache_file_path();
+        let file = File::create(file_path).expect("Unable to open autotune persistent cache file");
```

Comment:
`Unable to open autotune persistent cache file` => `Should be able to create the autotune persistent cache file`.